lib/Sema/CSBindings.cpp

```diff
@@ -2457,6 +2457,17 @@ TypeVariableBinding::fixForHole(ConstraintSystem &cs) const {
   unsigned defaultImpact = 1;
 
   if (auto *GP = TypeVar->getImpl().getGenericParameter()) {
+    // If there are any other argument mismatches already detected for this
+    // call, we increase the score even higher so more argument fixes means
+    // less viable is the overload.
+    if (llvm::any_of(cs.getFixes(), [&](const ConstraintFix *fix) {
+          auto *fixLocator = fix->getLocator();
+          return fixLocator->findLast<LocatorPathElt::ApplyArgToParam>()
+                     ? fixLocator->getAnchor() == dstLocator->getAnchor()
+                     : false;
+        }))
+      defaultImpact += 6;
```

Comment:
It would be good to figure out how to do this without having look for fixes because 
generally speaking a generic parameter could be involved in multiple arguments or result positions, it might be not inferrable because i.e. there was a member or some other remote problem that propagated placeholders to arguments/result without recording a fix (just like in example with the issue).  I see a couple of ways here - either be more aggressive with argument mismatch fixes i.e. if there an argument mismatch and there are some type variables involved in the type - increase the impact of the mismatch fix or increase default impact of the `DefaultGenericArgument` fix.

Reply:
Increasing default impact of `DefaultGenericArgument` fix in order to make this particular issue give a good diagnosis would require a big number, which is likely to affect a lot. 

Increasing impact of an argument mismatch with a type containing type variables seems better, and is an approach I looked at doing here - I'll go back and try that.